iss/app.py

```diff
@@ -28,292 +32,103 @@ def generate_reads(args):
-                except AssertionError:
-                    logger.error(
-                        "--ncbi and --n_genomes_ncbi of unequal lengths. \
-                        Aborting"
-                    )
-                    sys.exit(1)
-                for g, n in zip(*args.ncbi, *args.n_genomes_ncbi):
-                    genomes_ncbi = download.ncbi(g, n, args.output + "_ncbi_genomes.fasta")
-                genome_files.append(genomes_ncbi)
-            if args.ncbi and not args.n_genomes_ncbi:
-                logger.error("--ncbi/-k requires --n_genomes_ncbi/-U. Aborting.")
-                sys.exit(1)
+    readcount_dic, abundance_dic = load_readcount_or_abundance(
+        args.readcount_file,
+        args.abundance_file,
+        args.coverage_file,
+        args.coverage,
+        args.abundance,
+        args.draft,
+        genome_list,
+        genome_file,
+        args.n_reads,
+        args.output,
+        error_model,
+    )
 
-        else:
-            logger.error("One of --genomes/-g, --draft, --ncbi/-k is required")
-            sys.exit(1)
+    logger.info("Using %s cpus for read generation" % args.cpus)
 
-        genome_file = args.output + ".iss.tmp.genomes.fasta"
-        util.concatenate(genome_files, output=genome_file)
+    if readcount_dic is not None:
+        n_reads = sum(readcount_dic.values())
+        logger.info("Generating %s reads" % n_reads)
+    elif not (args.coverage or args.coverage_file):
+        n_reads = util.convert_n_reads(args.n_reads)
+        logger.info("Generating %s reads" % n_reads)
 
-        # for n_genomes we use reservoir sampling to draw random genomes
-        # from the concatenated genome file. We then override the file.
-        if args.n_genomes and not args.draft and not args.ncbi:
-            genome_count = util.count_records(genome_file)
-            genome_files = [
-                genome for genome in util.reservoir(SeqIO.parse(genome_file, "fasta"), genome_count, args.n_genomes)
-            ]
-            SeqIO.write(genome_files, genome_file, "fasta")
+    try:
+        # list holding the prefix for each cpu's temp file
+        temp_file_list = [f"{args.output}.iss.tmp.{i}" for i in range(args.cpus)]
+        f = open(genome_file, "r")  # re-opens the file
+        with f:
+            fasta_file = SeqIO.parse(f, "fasta")
+            # TODO check if a SeqIO.index (db) leads to better memory usage and is not slower
+            # fasta_dict = SeqIO.index(f, 'fasta')
+
+            # Calculate how many reads we want each cpu to generate
+            n_read_pairs = n_reads // 2
+            chunk_size = -((n_read_pairs) // -args.cpus)  # this is ceildiv, see https://stackoverflow.com/a/17511341
+            logger.debug("Chunk size: %s" % chunk_size)
+
+            # Divide the work of generating n_reads for each record into chunks
+            work_chunks = list(
+                generate_work_divider(
```

Comment:
why does this function `yield` chunks? you seem to be consuming the whole iterator and storing it in a list directly. Or am I missing something?

Reply:
Ah, yes the `list()` should be removed here. The reason for it being a generator is mostly because it was convienient to get the logic working.